File: backend/app/models/referral.py
from datetime import datetime, timezone
from decimal import Decimal
from sqlalchemy import DateTime, ForeignKey, Numeric, String, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column
from app.core.database import Base

class Referral(Base):
    __tablename__ = "referrals"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    referrer_id: Mapped[int] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    referred_user_id: Mapped[int] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), nullable=False, unique=True, index=True)
    referral_code: Mapped[str] = mapped_column(String(20), nullable=False, index=True)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="PENDING", index=True)
    reward_amount: Mapped[Decimal] = mapped_column(Numeric(14, 2), nullable=False, default=Decimal("200.00"))
    reward_transaction_id: Mapped[int | None] = mapped_column(ForeignKey("wallet_transactions.id", ondelete="SET NULL"), nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), nullable=False)

    __table_args__ = (
        UniqueConstraint("referrer_id", "referred_user_id", name="uq_referral_referrer_referred"),
    )

File: frontend/services/referral.ts
import { api } from "@/lib/api";

export interface Referral {
  id: number;
  referrer_id: number;
  referred_user_id: number;
  referral_code: string;
  status: string;
  reward_amount: number | string;
  completed_at?: string | null;
  created_at: string;
}

export async function getMyReferralCode() {
  const { data } = await api.get<{ referral_code: string }>("/referrals/my-code");
  return data;
}

export async function getReferralHistory() {
  const { data } = await api.get<Referral[]>("/referrals/history");
  return data;
}

export async function applyReferralCode(referral_code: string) {
  const { data } = await api.post<Referral>("/referrals/apply", { referral_code });
  return data;
}
